### PR TITLE
WIP: cwltoil.py refactors for conditionals

### PR DESCRIPTION
**Pull request by kaushik-work**
Some changes are general refactoring of code. Some changes are additions to allow for the conditionals proposal.

The code does not affect existing functionality, so this code will run CWL 1.0 documents just fine.

### CODE REVIEW

**Comment by DailyDreaming** [comment]
@kaushik-work Thanks for this work!  Will try to review tomorrow or Thursday.

**Comment by mr-c** [comment]
@DailyDreaming You can review for fun, but this PR is more for @kaushik-work and I to see what he's done so far in his sketch of a future CWL feature.

**Comment by kaushik-work** [comment]
@DailyDreaming thanks for taking a look! This is a WIP for implementing conditionals in CWL + some refactoring of the cwltoil code. Hope to ping you soon when it is in a state where it should be considered for merging. The changes will be isolated to this one file (or at least under the `cwl` directory)

**Comment by DailyDreaming** [comment]
@mr-c I see.  You have more experience on the CWL end so I'll leave it to you, but as a dev here I still feel obligated to look over new PRs in case of breaking changes.

**Comment by DailyDreaming** [comment]
@kaushik-work Understood.  Thank you and @mr-c for propagating the changes here.

File: src/toil/cwl/cwltoil.py [hunk 1]
```diff
@@ -985,21 +1087,14 @@ def run(self, file_store):
...
-                if link_merge == "merge_nested":
-                    outobj[key] = (
-                        MergeInputsNested(
-                            [(shortname(s), promises[s].rv())
-                             for s in aslist(out["outputSource"])]))
-                elif link_merge == "merge_flattened":
-                    outobj[key] = (
-                        MergeInputsFlattened([
-                            (shortname(s), promises[s].rv())
-                            for s in aslist(out["source"])]))
```
Comment (kaushik-work) [inline src/toil/cwl/cwltoil.py]: FYI @tetron, @mr-c  this is definitely a bug - and the new code fixes this.

**Comment by DailyDreaming** [comment]
@kaushik-work What is the status on this?

**Comment by kaushik-work** [comment]
Hi @DailyDreaming . I'm on vacation right now, will be back in a week or so. 

From my point of view it is complete. @tetron asked for an additional feature (an `else`-like clause), but after thinking about it some more, I will suggest we go with this for CWL 1.2 and defer any additions until this feature has stabilized and we've gotten usage feedback from the community.

I will post an update here after we've discussed this. 

Thanks!

**Pull request closed without merging**